### DISCUSSION TITLE
Debugging: what does `--to native` do / seeing pandoc AST after filters

### DISCUSSION THREAD

**Discussion by GuillaumeDehaene** [quarto-dev/quarto-cli#7285] [Q&A] [answered]
### Description

Hello everyone,

My current understanding of the Quarto pipepline is:

1. Execute code and replace it with quarto markdown
2. Parse quarto markdown with pandoc, producing an initial pandoc AST document
3. Execute some quarto filters on the AST, producing a final pandoc AST document
    This is the step where syntax that is specific to quarto markdown is interpreted
    Is the filter somehow keeping in memory information from the original markdown document?
5. Use a Pandoc writer to convert the final AST to the desired output.
    The writer can be customized, for example by changing the templates.

I would like to see to the final AST document, both for debugging purposes and to satisfy my curiosity.
Is that actually what `quarto render source.qmd --to native` produces?

I'm unsure because the following MWE shows no difference in the outputted AST between a named codeblock and unnamed codeblock:

source.qmd
`````
```{filename="FILENAME"}
CONTENTA
```

```
CONTENTB
```
`````

source.native, produced by `quarto render source.qmd --to native` (quarto version 1.3.450)
`````
Pandoc
  Meta { unMeta = fromList [] }
  [ CodeBlock ( "" , [] , [] ) "CONTENTA"
  , CodeBlock ( "" , [] , [] ) "CONTENTB"
  ]
`````

**Comment by cderv** [comment, marked as answer]
Which version of Quarto are you using ? 

We may have (fix / improve) as with latest 1.4, I get the information. 

````
> quarto render index.qmd --to native -o -
pandoc -o C:\Users\chris\AppData\Local\Temp\quarto-session29e06f12\541b9f25\90fd9712.native
  to: native
  standalone: true
  default-image-extension: png

metadata
  title: oops

Pandoc
  Meta
    { unMeta =
        fromList [ ( "title" , MetaInlines [ Str "oops" ] ) ]
    }
  [ CodeBlock
      ( "" , [] , [ ( "filename" , "FILENAME" ) ] ) "CONTENTA"
  , CodeBlock ( "" , [] , [] ) "CONTENTB"
  ]
````

@cscheid is the new `at` step for filters impacting the AST ? I have a doubt if format native is the very last AST before writers ? 

thanks for your insights!

**Reply by GuillaumeDehaene** [reply]
Thank you for your answer.

That was on version 1.3450
I get the same output as you on 1.4.429
At least on this simple example, it seems like the AST indeed corresponds to the final version.

**Reply by cscheid** [reply]
The `--to native` output will always correspond to the result of running the entire Lua pipeline.

> @cscheid is the new at step for filters impacting the AST ?

No, they're unrelated features.

> I have a doubt if format native is the very last AST before writers ?

`--to native` is not a filter option, it's a Pandoc option, so our filters all execute. They might execute differently because we have format-specific checks, though.
